src/layers/learning/fully_connected.cpp

```
@@ -210,4 +231,160 @@ void fully_connected_layer<data_layout::DATA_PARALLEL>::bp_compute_cpu() {
+                 gradient_wrt_output.LockedBuffer(), gradient_wrt_output_ldim,
+                 ones_d.get_data(0), 1,
+                 DataType(0),
+                 m_bias_gradient->Buffer(), 1);
+    bias_optimizer->add_to_gradient_staging(
+                                            *m_bias_gradient,
+                                            m_bias_scaling_factor / this->m_model->get_effective_mini_batch_size());
+  }
+
+  // Compute gradient w.r.t. linearity if needed
+  optimizer* linearity_optimizer = this->m_weights[0]->get_optimizer();
+  if (linearity_optimizer != nullptr) {
+    CHECK_CUDA(cudaSetDevice(this->m_cudnn->get_gpu()));
+    cublas::gemm(this->m_cudnn->get_cublas_handle(),
+                 CUBLAS_OP_N, CUBLAS_OP_T,
+                 output_size, input_size, mini_batch_size,
+                 DataType(1),
+                 gradient_wrt_output.LockedBuffer(), gradient_wrt_output_ldim,
+                 input.LockedBuffer(), input_ldim,
+                 DataType(0),
+                 m_linearity_gradient->Buffer(), output_size);
+    linearity_optimizer->add_to_gradient_staging(
+                                                 *m_linearity_gradient,
+                                                 DataType(1) / this->m_model->get_effective_mini_batch_size());
+  }
+
+  // Compute gradient w.r.t. input
+  if (mini_batch_size != 0) {
+    CHECK_CUDA(cudaSetDevice(this->m_cudnn->get_gpu()));
+    cublas::gemm(this->m_cudnn->get_cublas_handle(),
+                 CUBLAS_OP_T, CUBLAS_OP_N,
+                 input_size, mini_batch_size, output_size,
+                 DataType(1),
+                 linearity, output_size,
+                 gradient_wrt_output.LockedBuffer(), gradient_wrt_output_ldim,
+                 DataType(1),
+                 gradient_wrt_input.Buffer(), gradient_wrt_input_ldim);
+  }
+
+#endif // LBANN_HAS_CUDNN
+}
+
+template <>
+void fully_connected_layer<data_layout::MODEL_PARALLEL, El::Device::GPU>::fp_compute() {
```

We should throw an error here.

---

I have implemented this now.